protocols/shell-foreign-toplevel-icon-v1.xml

```diff
@@ -0,0 +1,271 @@
+      <entry name="destroyed_while_unfinished"
+             summary="attempted to destroy an icon which has not received a finished event"
+             value="2"/>
+    </enum>
+
+    <request name="destroy">
+      <description summary="destroys the icon">
+        This request destroys the icon object. Destroying the icon
+        prior to receiving a 'finished' event is a protocol error.
+      </description>
+    </request>
+
+    <event name="finished">
+      <description summary="the compositor is finished with this icon">
+        This event indicates that this icon will not be used in further
+        'shell_foreign_toplevel_icon_handle_v1.icon' events and is safe
+        to destroy.
+
+        This event may be sent in response to a release request or
+        when no toplevel icon handles refer to this icon anymore.
+      </description>
+    </event>
+
+    <request name="release">
+      <description summary="attempt to release the icon">
+        This request indicates that the client would like to destroy
+        the icon object.
+
+        The compositor must respond by either sending a
+        'shell_foreign_toplevel_icon_handle_v1.icon' event referencing
+        this icon, or sending the 'finished' event to this object.
+      </description>
+    </request>
+
+    <request name="copy_icon">
+      <description summary="copy a shm backed icon">
+        Copy an icon listed by 'buffer' into a client provided buffer.
+
+        The provided size and format must have been listed by 'buffer'
+        in the last icon option group.
+      </description>
+      
+      <arg name="size" type="uint"/>
+      <arg name="format" type="uint" enum="wl_shm.format"/>
+      <arg name="buffer" type="object" interface="wl_buffer"/>
+    </request>
+
+    <event name="done">
+      <description summary="all icon representation have been sent">
+        This event is sent after all 'add_buffer' events and optionally
+        an 'icon_name' event have been sent.
+      </description>
+    </event>
+
+    <event name="icon_name">
+      <description summary="set an icon name">
+        This event assigns an icon name to this icon. It must not be
+        sent more than once, and must not be sent after 'done'.
+
+        The icon refered to by the provided name should be resolved
+        according to the XDG icon theme specification[1] using the
+        environment's current icon theme.
+
+        [1]: https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html
+      </description>
+
+      <arg name="name" type="string"/>
+    </event>
+
+    <event name="add_buffer">
+      <description summary="provides an icon buffer">
+        This event exposes an available pixel buffer representing the icon.
+        Multiple add_buffer events may be sent at different sizes, scales,
+        and buffer formats. The event must not be sent after 'done'.
+
+        The buffer must be backed by wl_shm and must be square.
```

Comment:
Is this enforced somewhere else in icon specs or protocol? Idk but seems a bit strange a requirement